### DISCUSSION TITLE
Implementing Redirects

### DISCUSSION THREAD

**Discussion by andy5995** [Lyrics/lyrics#521] [Ideas]
I suppose it's not very likely that there's a good way at present to implement redirects, but thought I'd kick off a discussion about it.

['.htaccess' files](https://help.dreamhost.com/hc/en-us/articles/215747748-How-can-I-redirect-and-rewrite-my-URLs-with-an-htaccess-file-)  aren't an option.

There's a [jekyll-redirect-from plugin](https://github.com/jekyll/jekyll-redirect-from) that uses the 'HTTP-REFRESH' meta tag. This site isn't generated with jekyll, but I wonder if there was a case where we wanted to delete a file, if it would be feasible to just remove it and replace it with a small file that just contained the refresh meta tag to redirect where the new file is.

Other than that, I have no ideas atm.

Well.. publishing the site to a regular webhost is probably an option. Then .htaccess files could get published to the server. I don't know if that can be done from GitHub. It could be done if there's a git repo on the webhost server itself

https://stackoverflow.com/questions/17287517/sync-github-repo-to-dreamhost-site

Of course that involves a lot overhead and cost so probably not worth it.

**Comment by snshn** [comment]
So far [this](https://github.com/Lyrics/lyrics-website-generator/blob/master/src/templates/404.mustache#L5) is the only form of redirects I managed to add, and it requires JS to be functioning.

I assume the database could be parsed and converted into SQL or a very big JSON file, then used to power some dynamic website. lyrics.github.io is a lightweight web UI more than a fully-fledged website with built-in search, I wouldn't bet too much on its capabilities... not to mention monthly traffic limit that GitHub imposes on GitHub Pages websites.

What, do you think, would best represent use-cases for implementing redirects?

**Reply by andy5995** [reply]
> So far [this](https://github.com/Lyrics/lyrics-website-generator/blob/master/src/templates/404.mustache#L5) is the only form of redirects I managed to add, and it requires JS to be functioning.

```
Redirect to https://example.com/ after 5 seconds:

<meta http-equiv="refresh" content="5; url=https://example.com/">

Redirect to https://example.com/ immediately:

<meta http-equiv="refresh" content="0; url=https://example.com/">
```

> 
> I assume the database could be parsed and converted into SQL or a very big JSON file, then used to power some dynamic website. lyrics.github.io is a lightweight web UI more than a fully-fledged website with built-in search, I wouldn't bet too much on its capabilities... not to mention monthly traffic limit that GitHub imposes on GitHub Pages websites.
> 
> What, do you think, would best represent use-cases for implementing redirects?

If there was a duplicate set of lyrics that we found completely unnecessary. I don't see a point to having duplicates for singles, albums, etc. I only know of one current case "Butterfly" by Nicole. But it would be nice to delete the version from the single release directory. If an edit ever needs to be done, it needs to happen to both.

Would be nice if one could just be replaced with a small file that could use the refresh thing (above).

Probably to remove duplicates would be the main reason. But I don't know if this project has many of them, or specific policy about them.

**Comment by snshn** [comment]
Duplicates do happen... can't say how often, but singles rarely contain unique text or different version of the song, so most of the time there's no point for singles to be in the database — as like you've said, they're unnecessary.

Redirects make sense in case things get moved, but realistically, it'll be more than just a couple of lines of code, and more code will make lyrics-website-generator even more complicated. I'll need to make unit tests for it, split into modules, and then figure out the final plan for redirects.

We could make a metadata property `Copy of` or `Duplicate of`, that would let us run a unit test to compare two texts (including `Original text by` key).

There's no policy really, the project is evolving and getting better thanks to contributors like yourself. As long as we use logic we'll find a good solution.
